### PR TITLE
Adds Quirk Customization

### PR DESCRIPTION
**Pull request by That0nePerson**
## About The Pull Request
Adds the ability to customize your character's quirks after selecting them. This includes the location of your prosthetic limb, the type of drug for junkie, the language for multilingual, the type of phobia, and the drink/smoke type you spawn with for alcoholic/smoker.

These are the currently available languages for multilingual, tell me if I should remove or add some. I tried to base it off of what the multilingual trait could give before, but I may have missed some.
image

I have no idea what I'm doing when it comes to the schema and database, corrections are veyr much appreciated.

## Why It's Good For The Game
Trying to work quirks into your character's backstory was very difficult before, but now it should be a lot easier and encourage people to come up with stories for how they lost a limb or learned a language.

## Testing Photographs and Procedure

<details>
<summary>Screenshots&Videos</summary>

image
I started the game twice with each of the quirks to see if they worked, and tested with debug text to see if it was just defaulting to the random choice and I got lucky or if it respected my choice in character creation. Let me know if I should do more.

</details>

## Changelog
:cl:
add: You can now customize the prosthetic limb, phobia, junkie, multilingual, alcoholic, and smoker quirks!
/:cl:

### CODE REVIEW

**Comment by HowToLoLu** [comment]
There should probably still be the ability to let the game choose the random option for you in addition to this.

File: SQL/database_changelog.txt
Comment (itsmeow) [inline SQL/database_changelog.txt]: These should include `AFTER` clauses, so they are inserted at a sensible location order-wise in the table schema (as defined in beestation_schema.sql).

So 

```
ADD COLUMN `quirk_prosthetic_limb_location` VARCHAR(64) COLLATE 'utf8mb4_general_ci' AFTER `all_quirks`,
```

and so on

File: SQL/database_changelog.txt
Comment (itsmeow) [inline SQL/database_changelog.txt:5]: these values are also defined in code somewhere, so need to be updated there

File: code/__DEFINES/language.dm
Comment (itsmeow) [inline code/__DEFINES/language.dm]: this should probably just be a global list rather than a define

`GLOBAL_INIT_VAR(name, typecacheof(list(...)))`

File: code/datums/traits/negative_quirk.dm
Comment (itsmeow) [inline code/datums/traits/negative_quirk.dm]: does `mind.current.client` not work?  or `quirk_target.client`?

Also put `ckey(quirk_holder.key)` into a var before passing it to directory

File: code/datums/traits/negative_quirk.dm
Comment (itsmeow) [inline code/datums/traits/negative_quirk.dm]: `prefs` can be null, check `if(C?.prefs)` instead

File: code/modules/client/preferences/entries/character/quirkoptions.dm
Comment (itsmeow) [inline code/modules/client/preferences/entries/character/quirkoptions.dm]: these can be randomized. this just determines if random body should randomize it, all of these, yes, that's fine

File: code/modules/client/preferences/entries/character/quirkoptions.dm
Comment (itsmeow) [inline code/modules/client/preferences/entries/character/quirkoptions.dm]: you can move the init order of SStraumas higher, this should also be a DEFINE or global list otherwise. you should never rely on copypasting something from another source as someone could forget to update it

File: code/modules/client/preferences/entries/character/quirkoptions.dm
Comment (itsmeow) [inline code/modules/client/preferences/entries/character/quirkoptions.dm]: Make a generic subtype of `/datum/preference/choiced` for quirks, have a var on it called `required_quirk_name` that checks this, then you can just set that on the parent. It can also implement `apply_to_human` as nothing to avoid the boilerplate here.

File: code/modules/client/preferences/entries/character/quirkoptions.dm
Comment (itsmeow) [inline code/modules/client/preferences/entries/character/quirkoptions.dm]: Yeah a global list here is waaay better rather than a define

**Review by itsmeow — CHANGES_REQUESTED** [review]

**Comment by itsmeow** [comment]
I do plan to make UI changes to more sensibly organize where quirk preferences go, but this is a very functional basis that can be adjusted to match that in the future. I specifically requested that it be implemented this way for the time being as I don't have time to do it myself at the moment.

It is also possible to let the game randomize this for you with the new randomization system, although it's not very clear. You can enable body randomization which allows you to specify which preferences you specifically want to be randomized.

File: code/datums/traits/negative_quirk.dm
Comment (That0nePerson) [reply code/datums/traits/negative_quirk.dm]: both of those are null at the time of on_spawn(), I tried using those first

File: SQL/database_changelog.txt
Comment (That0nePerson) [reply SQL/database_changelog.txt:5]: updated

File: code/__DEFINES/language.dm
Comment (That0nePerson) [reply code/__DEFINES/language.dm]: done !

File: code/datums/traits/negative_quirk.dm
Comment (That0nePerson) [reply code/datums/traits/negative_quirk.dm]: put it into a var :3

File: SQL/database_changelog.txt
Comment (That0nePerson) [reply SQL/database_changelog.txt]: added! ty for the help

File: code/datums/traits/negative_quirk.dm
Comment (That0nePerson) [reply code/datums/traits/negative_quirk.dm]: added

File: code/modules/client/preferences/entries/character/quirkoptions.dm
Comment (That0nePerson) [reply code/modules/client/preferences/entries/character/quirkoptions.dm]: couldn't get moving the init order to work, it's a define now

File: code/modules/client/preferences/entries/character/quirkoptions.dm
Comment (itsmeow) [inline code/modules/client/preferences/entries/character/quirkoptions.dm]: if you use `///` triple slash comments on the line before it has better VSCode support as hovering the var name will show the comment

File: code/modules/client/preferences/entries/character/quirkoptions.dm
Comment (itsmeow) [inline code/modules/client/preferences/entries/character/quirkoptions.dm]: this comment is wrong now, remove it

File: SQL/beestation_schema.sql
Comment (itsmeow) [inline SQL/beestation_schema.sql:499]: need to update `DB_MINOR_VERSION` in `code/__DEFINES/subsystems.dm`

It is infuriating how many places this is in lol. that's the last one

**Review by itsmeow — CHANGES_REQUESTED** [review]

**Comment by EvilDragonfiend** [comment]
can you multiselect prosthetic limbs?
I think such button type should be multi-selectable by a maximum selection value.

**Comment by That0nePerson** [comment]
@EvilDragonfiend I really want that and I would do that, but it would require me to make a lot of changes to TGUI which I have absolutely no experience with

**Comment by EvilDragonfiend** [comment]
> @EvilDragonfiend I really want that and I would do that, but it would require me to make a lot of changes to TGUI which I have absolutely no experience with

my request is more tied to how `/datum/preference` works rather than TGUI. It seems it's better to revisit someday than now.

File: SQL/database_changelog.txt
Comment (PowerfulBacon) [inline SQL/database_changelog.txt]: These need checks so that if run multiple times, it will work as intended (See the IF NOT EXISTS) clause of other SQL scripts.

Comment (That0nePerson) [reply SQL/database_changelog.txt]: thank you, done

**Comment by itsmeow** [comment]
we don't have a list data type nor is there really an easy way to implement it without designing something new. so i asked for choiced types for the time being

**Review by Crossedfall — COMMENTED** [review]
My interpretation of the changes and how we're setting up future additions worries me a bit. We're adding in six new columns into an already pretty large table with the expectation that future quirk customizations would need to do the same. I'm not really sold on this being the ideal implementation. It certainly works, but I don't think it's setting up very good practices in the future and has a reasonably high chance at ballooning into something that's pretty difficult to maintain or switch away from in the long-term.

**Comment by itsmeow** [comment]
The alternatives are making a new long table vs flat tied per character which makes some goofy data integrity problems, or adding another big JSON field which removes the datumization ability entirely and is a step backwards

**Comment by PowerfulBacon** [comment]
I do agree with crossed, since getting a new column every time we want a new pref could be a bit of a pain. Having a table that is ckey, character slot, preference key and value seems to make sense to me since then we only need to populate it if changed from the default.

**Comment by itsmeow** [comment]
This is why I wanted that in the first place 😵‍💫

**Comment by That0nePerson** [comment]
I have no idea how the database works, I can just close this and let someone who knows how to do this do it if this would cause future issues or something

**Comment by itsmeow** [comment]
See #9899

**Comment by Crossedfall** [comment]
OK. So, we've had a number of discussions about this internally and while I still think that the potential for a problem still exists... For the time being, I'm OK with moving forward with the current proposition. 

We can adjust things later on if needed without too much additional effort.

File: beestation.dme
Comment (itsmeow) [inline beestation.dme:2235]: This file should follow standard naming conventions. Be sure to rename the actual file as well.
```suggestion
#include "code\modules\client\preferences\entries\character\quirk_options.dm"
```

File: code/datums/traits/positive_quirk.dm
Comment (itsmeow) [inline code/datums/traits/positive_quirk.dm:145]: do keep in mind that prefs don't actually ever come out as null, you'd need to specify a special preference named "Random" and make it the default, as well as disable randomization for this preference to make it random in this way.

**Review by itsmeow — CHANGES_REQUESTED** [review]

**Comment by EvilDragonfiend** [comment]
https://github.com/BeeStation/BeeStation-Hornet/pull/10021

This is almost done. You'll have a sane place to display quirk options once it's merged.

**Comment by Rukofamicom** [comment]
Good PR or not, it's been a month

**Comment by EvilDragonfiend** [comment]
No, no, it's basically on hold because this PR is not yet good to accept due to how pref works. Kat and I are fixing it.

**Comment by itsmeow** [comment]
No we can merge this as is, I'd just like my changes addressed.

**Pull request closed without merging**
